src/cargo/core/registry.rs

```diff
@@ -175,6 +180,41 @@ impl<'cfg> PackageRegistry<'cfg> {
         sub_vec.push((id, deps));
     }
 
+    pub fn augment(&mut self,
+                   url: &Url,
+                   deps: &[Dependency]) -> CargoResult<()> {
+        let deps = deps.iter().map(|dep| {
+            let mut summaries = self.query_vec(dep)?.into_iter();
+            let summary = match summaries.next() {
+                Some(summary) => summary,
+                None => {
+                    bail!("augment for `{}` in `{}` did not resolve to any crates",
+                          dep.name(), url)
+                }
+            };
+            if summaries.next().is_some() {
+                bail!("augment for `{}` in `{}` resolved to more than one candidate",
+                      dep.name(), url)
+            }
+            if summary.package_id().source_id().url() == url {
+                bail!("augment for `{}` in `{}` points to the same source, but \
+                       augmentations must point to different sources",
+                      dep.name(), url);
+            }
+            Ok(summary)
+        }).collect::<CargoResult<Vec<_>>>().chain_err(|| {
+            format!("failed to resolve augments for `{}`", url)
+        })?;
+
+        *self.augments.entry(url.clone()).or_insert(Vec::new()) = deps;
```

Comment:
I feel like I originally had this as `extend` and later changed it to this. I'll update the code.